SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriverRPC.cpp

```diff
@@ -0,0 +1,471 @@
+	}
+}
+
+USpatialNetDriverRPC::StandardQueue::SentRPCCallback USpatialNetDriverRPC::MakeRPCSentCallback(TArray<UpdateToSend>& OutUpdates)
+{
+	if (EventTracer != nullptr)
+	{
+		return [&OutUpdates, this](FName Name, Worker_EntityId EntityId, Worker_ComponentId ComponentId, uint64 RPCId,
+								   const FSpatialGDKSpanId& SpanId) {
+			return OnRPCSent(EventTracer, OutUpdates, Name, EntityId, ComponentId, RPCId, SpanId);
+		};
+	}
+	return StandardQueue::SentRPCCallback();
+}
+
+RPCCallbacks::DataWritten USpatialNetDriverRPC::MakeDataWriteCallback(TArray<FWorkerComponentData>& OutArray) const
+{
+	return [&OutArray](Worker_EntityId EntityId, Worker_ComponentId ComponentId, Schema_ComponentData* InData) {
+		return OnDataWritten(OutArray, EntityId, ComponentId, InData);
+	};
+}
+
+RPCCallbacks::UpdateWritten USpatialNetDriverRPC::MakeUpdateWriteCallback(TArray<UpdateToSend>& OutUpdates) const
+{
+	return [&OutUpdates](Worker_EntityId EntityId, Worker_ComponentId ComponentId, Schema_ComponentUpdate* InUpdate) {
+		return OnUpdateWritten(OutUpdates, EntityId, ComponentId, InUpdate);
+	};
+}
+
+void USpatialNetDriverRPC::Init(USpatialNetDriver* InNetDriver, const SpatialGDK::FSubView& InActorAuthSubView,
+								const SpatialGDK::FSubView& InActorNonAuthSubView)
+{
+	NetDriver = InNetDriver;
+	LatencyTracer = USpatialLatencyTracer::GetTracer(InNetDriver->GetWorld());
+	EventTracer = InNetDriver->Connection->GetEventTracer();
+	RPCService = MakeUnique<SpatialGDK::RPCService>(InActorNonAuthSubView, InActorAuthSubView);
+
+	{
+		// TODO
+		// MulticastReceiver =
+	}
+}
+
+void USpatialNetDriverRPC::AdvanceView()
+{
+	RPCService->AdvanceView();
+}
+
+void USpatialNetDriverRPC::ProcessReceivedRPCs()
+{
+	// MulticastReceiver->
+}
+
+TArray<FWorkerComponentData> USpatialNetDriverRPC::GetRPCComponentsOnEntityCreation(const Worker_EntityId EntityId)
+{
+	static TArray<Worker_ComponentId> EndpointComponentIds = {
+		SpatialConstants::CLIENT_ENDPOINT_COMPONENT_ID, SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID,
+		/*SpatialConstants::MULTICAST_RPCS_COMPONENT_ID,
+		SpatialConstants::CROSSSERVER_SENDER_ENDPOINT_COMPONENT_ID*/
+	};
+
+	TArray<FWorkerComponentData> Components;
+	GetRPCComponentsOnEntityCreation(EntityId, Components);
+
+	for (auto EndpointId : EndpointComponentIds)
+	{
+		if (Components.FindByPredicate([EndpointId](const FWorkerComponentData& Data) {
+				return Data.component_id == EndpointId;
+			})
+			== nullptr)
+		{
+			FWorkerComponentData Data;
+			Data.component_id = EndpointId;
+			Data.schema_type = Schema_CreateComponentData();
+			Components.Add(Data);
+		}
+	}
+
+	return Components;
+}
+
+void USpatialNetDriverRPC::GetRPCComponentsOnEntityCreation(const Worker_EntityId EntityId, TArray<FWorkerComponentData>& OutData) {}
+
+void USpatialNetDriverRPC::FlushRPCUpdates()
+{
+	const TMap<FName, RPCService::RPCReceiverDescription>& Receivers = RPCService->GetReceivers();
+	TArray<UpdateToSend> Updates;
```

Comment:
nit - 'Reserve()' for efficiency?

Reply:
I do not know which amount to reserve. Could be 0.

Reply:
That is okay?

Reply:
I believe it is, it's like polling for changes.
Sure, we could be smarter and accumulate counters everywhere, but that looks like optimization for a later time.

Reply:
Incremental array size memory allocs will always show up in a profile eventually, and it is trivial to fix them early.